website/docs/opengl/c1/2-hello-quad.md

```diff
@@ -0,0 +1,524 @@
+
+We've now successfully created and compiled our vertex shader. To compile our fragment shader, simply copy and paste the code you just wrote. Just make sure to change the `ShaderType` to `FragmentShader`!
+
+#### Creating the program
+Finally, we can create the **shader program**. In OpenGL, shaders are stored in a single **program** object.
+
+First, add the following to the top of your file:
+
+```cs
+private static uint _program;
+```
+
+Then, add the following to your `OnLoad` method:
+
+```cs
+_program = _gl.CreateProgram();
+```
+
+Now that we've created the program, we now need to attach the shaders to it and **link** them to the program.
+
+Add the following to your `OnLoad` method:
+
+```cs
+_gl.AttachShader(_program, vertexShader);
+_gl.AttachShader(_program, fragmentShader);
+
+_gl.LinkProgram(_program);
+
+_gl.GetProgram(_program, ProgramPropertyARB.LinkStatus, out int lStatus);
+if (lStatus != (int) GLEnum.True)
+    throw new Exception("Program failed to link: " + _gl.GetProgramInfoLog(_program));
+```
+
+Much like before, we must check to make sure that the program has linked correctly. Link errors are rare, most often caused by attempting to link with a shader that has failed to compile, which, assuming you implemented the checks earlier, should basically never happen.
+
+Now that we have linked the program, it can now function by itself, and we no longer need the shader objects. Therefore, we can **detach and delete** them, freeing a bit of GPU memory.
+
+Add the following to your `OnLoad` method:
+
+```cs
+_gl.DetachShader(_program, vertexShader);
+_gl.DetachShader(_program, fragmentShader);
+_gl.DeleteShader(vertexShader);
+_gl.DeleteShader(fragmentShader);
+```
+
+Now, run your program again. If you get no errors, this has all worked successfully!
+
+### Setting up the attributes
+The last thing we need to do before we can begin drawing our quad to the screen is to set up the attributes. By default, OpenGL does not know how to send the data to our shader, so we must tell it.
+
+<?# Warning "While fragment shaders (and other shaders) can have `in` attributes, the only ones you can directly set *outside* of a shader are the ones going into the vertex shader. The only way to set attributes in the fragment shader is to pass them through the vertex shader. Therefore, the attribute setup only affects the vertex shader and vertex buffer, not any other shader." /?>
+
+Add the following to your `OnLoad` method:
+
+```cs
+_gl.UseProgram(_program);
+
+const uint positionLoc = 0;
+_gl.EnableVertexAttribArray(positionLoc);
+_gl.VertexAttribPointer(positionLoc, 3, VertexAttribPointerType.Float, false, 3 * sizeof(float), (void*) 0);
```

Comment:
This code is related to the VAO. Readers need to understand that this code is configuring the VAO, and not the shader program.

This code is usually written together with the creation of the VAO.

Reply:
Hrm. I virtually always see this code right at the bottom of the vao setup, including in the official silk example. I didn't know that it was configuring specifically the VAO though, since I remember having issues with this if I didn't have a shader program bound, which made me assume it was configuring both.

(As a side note, personally, I absolutely despise VAOs and GL's attribute setup, I don't even use them outside of a single vao for the entire program. Give me input layouts any day of the week...)

Reply:
Yup, `EnableVertexAttribArray` and `VertexAttribPointer` apply config changes to the current VAO. Nothing to do with the current shader program.

The only thing that connects the VAO to the shader program are the attribute locations. In this case, it's just a single attribute, at location 0, of type vec3.

(What?? You use a single VAO for your entire program?? If you have to go around enabling and disabling attributes and calling `VertexAttribPointer` whenever you need to change vertex type or source VBOs, that's very inefficient, and you should use multiple VAOs)

Reply:
I do yeah, probably could use multiple VAOs though, I'll have to experiment with it.

I've changed the explanation to make it clear that its updating the VAO specifically (and also removed the bit that tells you to set the shader program)

> The last thing we need to do before we can begin drawing our quad to the screen is to set up the attributes. By default, OpenGL does not know how to send the data to our shader, so we must tell it.

> This is done with the **attribute setup**. This sets various parameters in the VAO, which tells OpenGL how to read the vertex data in the vertex buffer. As vertex buffers can contain more than just position data (it can contain almost anything you want), it is vital that OpenGL knows how to separate out the individual bits of data, so it can pass it to the shader correctly.

Hopefully this is okay?